### ISSUE TITLE
Windows: terminal freezes when a shiny app shuts down

### ISSUE THREAD

**Issue by daattali** [posit-dev/py-shiny#771]
I just installed py-shiny on a fresh Windows installation. I'm getting an issue where any time the shiny app is killed, the terminal freezes. If I run a shiny app with `shiny run`, then when I press ctrl+C I see messages saying that the shiny app is attempting to shut down, but the terminal gets stuck and never returns to the prompt.

If I run in `--reload` mode, then the moment I change a file and save it, the terminal gets to the same state. 

This happens in the command line and also in VSCode.

```
(venv) C:\Users\Dean\Documents\R\python\shinytest>shiny run
←[32mINFO←[0m:     Started server process [←[36m16176←[0m]
←[32mINFO←[0m:     Waiting for application startup.
←[32mINFO←[0m:     Application startup complete.
←[32mINFO←[0m:     Uvicorn running on ←[1mhttp://127.0.0.1:8000←[0m (Press CTRL+C to quit)
←[32mINFO←[0m:     127.0.0.1:49687 - "←[1mGET / HTTP/1.1←[0m" ←[32m200 OK←[0m
←[32mINFO←[0m:     ('127.0.0.1', 49689) - "WebSocket /websocket/" [accepted]
←[32mINFO←[0m:     connection open
←[32mINFO←[0m:     127.0.0.1:49688 - "←[1mGET / HTTP/1.1←[0m" ←[32m200 OK←[0m
←[32mINFO←[0m:     connection closed
←[32mINFO←[0m:     ('127.0.0.1', 49695) - "WebSocket /websocket/" [accepted]
←[32mINFO←[0m:     connection open
←[32mINFO←[0m:     Shutting down
```

**Comment by daattali** [comment]
(It would also be nice to fix https://github.com/posit-dev/py-shiny/issues/162 because this is a terrible experience on Windows)

**Comment by j-lim-sigma** [comment]
Hi, I experienced the same thing using python 3.12 on MacOS. It works fine with 3.11 though, both with --reload and without.

**Comment by jcheng5** [comment]
@j-lim-sigma Thank you, I'm installing 3.12 now to try to repro.

@daattali Any chance you're on Python 3.12 as well?

**Comment by daattali** [comment]
Yes I am indeed on 3.12. I can try to install 3.11 later if I find very simple instructions for it :)

**Comment by jcheng5** [comment]
I can repro on macOS 3.12 if I still have a connected browser. Once I close the browser, the command line returns. I get the same thing if I run the command `uvicorn app:app` instead of `shiny run`.

**Comment by jcheng5** [comment]
Here is a discussion on Uvicorn and a suggested fix, no response yet: https://github.com/encode/uvicorn/discussions/2122

**Comment by jcheng5** [comment]
FWIW, I tried daphne (`daphne app:app`) and hypercorn (`hypercorn app:app`) under Python 3.12 and they both served up my app fine and exited cleanly. So this appears to be specific to uvicorn.

**Comment by daattali** [comment]
> Here is a discussion on Uvicorn

Maybe it should be upgraded to an "Issue" in oder for the uvicorn team to notice?

**Comment by daattali** [comment]
I can confirm that downgrading to python 3.11 fixes this

**Comment by jcheng5** [comment]
This PR contains the fix. Thanks for your patience! https://github.com/encode/uvicorn/pull/2145

**Comment by corey-dawson** [comment]
Similar issue on python 3.10.13. Sometimes the process shuts down normally, sometimes the terminal freezes. for me. Have to close the terminal and reopen a new one. this is where mine gets stuck, but only maybe half the time:

![image](https://github.com/posit-dev/py-shiny/assets/79542680/4e9711b9-3fc4-4d4b-8f2f-095d2ea9155e)

**Comment by jcheng5** [comment]
Does it happen even with the simplest Shiny app?

```python
from shiny.express import input, render, ui

ui.input_slider("n", "N", 0, 100, 20)


@render.text
def txt():
    return f"n*2 is {input.n() * 2}"
```

**Comment by gza-austin-elliott** [comment]
Similar issue and symptoms as @corey-dawson above.

Windows 10.0.19045
Python 3.12

Fires up perfectly every time, but CTRL+C terminate does not consistently work. Occasionally requires full kill on the terminal to shut down. Something additional I haven't seen above which seems related is the terminal's outputs seem to freeze up when it fires up in this bugged mode. That is to say, I can tell when I'll have to kill the terminal rather than CTRL+C to shut down because the terminal will not display simple debug statements onstart. 

Ex
`# server logic
def server(input: Inputs, output: Outputs, session: Session):
    session_data = {}
    tables_trigger = reactive.Value(None) #init triggers
    data_trigger = reactive.Value(None) #init triggers
    ...[MORE CODE]...
    print("debug: export_button function defined")  # debug
    `
In a non-bugged startup this print statement displays in my terminal and the processes properly exit on CTRL+C. In a bugged startup this print statement does not display and the processes do not exit on CTRL+C.

**Comment by gza-austin-elliott** [comment]
> Does it happen with the simple app I posted in my previous comment?

It does not seem to. Did a fresh venv + your app and it initializes/terminates properly every time.

The app I was initially encountering this issue with was built on shiny.core. I quickly changed your example code over to shiny.core and tested and that also checked out, terminated fine with CRTL+C every time.

**Comment by jcheng5** [comment]
I don’t suppose you can share your source code with us? Either publicly or privately? If not, trying to find a reproducible example you can share would be a huge help.

**Comment by gza-austin-elliott** [comment]
I jumped back to my original source code to prep it to share and in retesting I could not recreate the issue. Very strange. I did not restart VSC, just changed from project to project. I'll keep an eye out and see if I can recreate it and pay closer attention to potential environmental variables.

**Comment by jcheng5** [comment]
@gza-austin-elliott I don't know if your Python version varies from project to project, but if so, I've noticed the Ctrl+C handling changed in Python 3.11+ in a way that would probably be highly relevant to this issue.
https://docs.python.org/3/library/asyncio-runner.html#handling-keyboard-interruption

Although the way they implementated that change, it is supposed to interrupt immediately if you hit Ctrl+C _twice_, which sounds like wasn't working for you, so maybe not...

**Comment by jcheng5** [comment]
If you still have a project where you can reproduce it, and are up for it, I'd be happy to jump on a zoom session with you and pair debug it. This is a scary problem for us to leave undiagnosed.

**Comment by jcheng5** [comment]
@ME-researchgroup Are you still on Python 3.10?

(On Python 3.11+) When asyncio is active, a single Ctrl+C makes a gentle suggestion to the current task that maybe it should interrupt at an opportune time. A second Ctrl+C is supposed to raise a KeyboardInterrupt on the main thread. Perhaps we should install our own Ctrl+C handler where if you do it _four_ times in quick succession, we terminate the process. Very unsatisfying but I hate the idea of people having to go to Task Manager.

**Comment by jcheng5** [comment]
Oh, wait, wait. So when you hit Ctrl+C it does show KeyboardInterrupt, but the process doesn't exit?

**Comment by jcheng5** [comment]
I’m at a loss. Are you able to reproduce this reliably enough that a live debugging session might be productive?
